Bid Viewability Module

pm-harshad-mane opened the pull request:
## Type of change
- [X] Feature

## Description of change
https://github.com/prebid/Prebid.js/issues/6205
- This module, when included, will trigger a BID_VIEWABLE event which can be consumed by Bidders and Analytics adapters
- GPT API is used to find when a bid is viewable, https://developers.google.com/publisher-tag/reference#googletag.events.impressionviewableevent
- The module does not work with adserver other than GAM with GPT integration
- Logic used to find a matching pbjs-bid for a GPT slot is ``` (slot.getAdUnitPath() === bid.adUnitCode || slot.getSlotElementId() === bid.adUnitCode) ``` this logic can be changed by using param ```customMatchFunction```
- When a rendered PBJS bid is viewable the module will trigger BID_VIEWABLE event, which can be consumed by bidders and analytics adapters
- For the viewable bid if ```bid.vurls type array``` param is and module config ``` firePixels: true ``` is set then the URLs mentioned in bid.vurls will be executed

# Params
- firePixels [optional] [type: boolean], when set to true, will fire the urls mentioned in bid.vurls which should be array of urls
- customMatchFunction [optional] [type: function(bid, slot)], when passed 

As both params are optional, you do not need to set config if you do not want to set value for any param

# Example of consuming BID_VIEWABLE event
```
	pbjs.onEvent('bidViewable', function(bid){
		console.log('got bid details in bidViewable event', bid);
	});

```

# Example of using config
```
	pbjs.setConfig({
        bidViewability: {
            firePixels: true,
            customMatchFunction: function(bid, slot){
                console.log('using custom match function....');
                return bid.adUnitCode === slot.getAdUnitPath();
            }
        }
    });
```

pm-harshad-mane commented:
Hello @jsnellbaker 
Can you please review this PR?

jsnellbaker commented on modules/bidViewability.js:
```diff
@@ -0,0 +1,52 @@
+
+        // if config is enabled AND VURL array is present then execute each pixel
+        if (globalModuleConfig[CONFIG_FIRE_PIXELS] === true && respectiveBid.hasOwnProperty(BID_VURL_ARRAY)) {
+          respectiveBid[BID_VURL_ARRAY].forEach(url => (new Image()).src = url);
```
Could you use the `utils.triggerPixel` function for loading the URL here?

pm-harshad-mane replied:
Thank you for the suggestion @jsnellbaker , I have updated the code now.

jsnellbaker reviewed (changes requested):
Hi @pm-harshad-mane 

I think overall this looks good.

Are there any concerns around GDPR consent permissions about loading the pixel URLs here?  I know there's the general on/off switch controlled by the publisher, but this module would still be loading a pixel URL by the bidder so that may run into the user's consent choices.

Could you also add some unit tests for the module?

pm-harshad-mane commented:
Thank you @jsnellbaker for reminding about GDPR, i have added GDPR and USP support , consent related params will be added to the URLs by the module. I have also added unit test cases.
@robertrmartinez can you please review?

pm-harshad-mane commented:
Hello @robertrmartinez ,
I have added documentation PR too, https://github.com/prebid/prebid.github.io/pull/2659

jsnellbaker reviewed (commented):
@pm-harshad-mane 

Thanks for making the updates.  To clarify though, what I meant before around consent referred more to the logic used in the `gdprEnforcement` in allowing/suppressing pixel URls.  eg https://github.com/prebid/Prebid.js/blob/62907bd6420f019a4548ad1294baff8f37613d71/modules/gdprEnforcement.js#L216

Should the viewable pixel URLs here adhere to that type of logic?

pm-harshad-mane commented:
Hello @jsnellbaker ,
The pixels we are trying to fire here will be for tracking the impression viewable event, like impression rendered event. It will not be related to user-sync, so I think we can fire these pixels (like analytics adapters) without GDPR Enforcement logic. There should not be a user related information in this pixel. Please correct me if my my understanding is inaccurate.
For now can we mention this limited scope in the documentation of the module and make changes later if some issues are raised?

bretg reviewed (changes requested):
@pm-harshad-mane - appreciate the module. Important question though -- it's not allowed for bidders to get notified about viewability events from other bidders. Just their own. Is that covered here? Tried to trace through the code but got lost down in events/_dispatch.

Would like for there to be a unit test confirming that a viewability event for a slot won by bidderA does not fire bidderB's listener.

bretg commented:
> Are there any concerns around GDPR consent permissions about loading the pixel URLs here?

Good line of questioning

My (non-lawyer) take is that as long as viewable notifications are constrained to only the bidder that won and displayed, hard to see any need for additional GDPR logic here. The auction already happened, the bidder won in the ad server, and was rendered. A viewability signal is pretty minor compared to all that stuff. Putting in a check to suppress this event seems way late -- like the auction shouldn't have happened at all.

So that's the main thing to confirm here - make sure only the winning bidder gets this signal.

pm-harshad-mane commented on modules/bidViewability.js:
```diff
@@ -0,0 +1,86 @@
+  return (slot.getAdUnitPath() === bid.adUnitCode || slot.getSlotElementId() === bid.adUnitCode);
+}
+
+export let getMatchingWinnigBidForGPTSlot = (globalModuleConfig, slot) => {
```
fixed the type.

pm-harshad-mane commented:
Thank you @bretg for your valuable inputs.
I need to work on emitting bid-viewable event to respective bidder only (among bidders), it will also be emitted to analytics adapters and for publisher to consume.

pm-harshad-mane commented:
Hello @bretg , @jsnellbaker 
Now i have added logic to trigger bidder.onBidViewable function, earlier i thought it is handle by events, my bad.
One question, when we want to trigger bid.someFunction then to find respective bidder should i refer to bid.bidder or bid.bidderCode ?

pm-harshad-mane commented:
Hello @robertrmartinez , 
Could you please review?

pm-harshad-mane commented:
Hello @robertrmartinez 
Can you please review this PR?

bretg commented:
@robertrmartinez isn't going to be able to get to this soon. Perhaps @musikele has the cycles?

pm-harshad-mane commented:
Hello @musikele 
Can you please review the changes?
@jsnellbaker, @bretg can you please mark "change requests" as resolved?

pm-harshad-mane commented:
Hello @musikele 
Could you please review the code?
@bretg can you please mark "change requests" as resolved?

bretg commented:
Note my review doesn't count as a code review.

mike-chowla commented on modules/bidViewability.js:
```diff
@@ -0,0 +1,88 @@
+    if (uspConsent) { queryParams.us_privacy = uspConsent; }
+
+    bid[BID_VURL_ARRAY].forEach(url => {
+      // we are assuming that "?" will be already present in the url
```
Is assuming '?' at the end of urls a good assumption?

pm-harshad-mane replied:
The assumption is made as these pixels will have additional params (for bid details) already present in URL, and then we need to add consent params. 
Most of the ad-tech companies use standard way of adding params in URL ( with `?` and `&` ), if there is going to be an exception then we will not be sure what delimiters are being used in such case.

pm-harshad-mane replied:
I have removed this assumption now, now the code will add the '?' if it is not present in URL

mike-chowla commented on modules/bidViewability.md:
````diff
@@ -0,0 +1,48 @@
+- Logic used to find a matching pbjs-bid for a GPT slot is ``` (slot.getAdUnitPath() === bid.adUnitCode || slot.getSlotElementId() === bid.adUnitCode) ``` this logic can be changed by using param ```customMatchFunction```
+- When a rendered PBJS bid is viewable the module will trigger BID_VIEWABLE event, which can be consumed by bidders and analytics adapters
+- For the viewable bid if ```bid.vurls type array``` param is and module config ``` firePixels: true ``` is set then the URLs mentioned in bid.vurls will be executed. Please note that GDPR and USP related parameters will be added to the given URLs, here we have assumed that URLs will always have "?" symbol included.
+
````
Is there a way for the bidder to tell if viewability module is present?  I ask because viewability stats aren't useful unless you know if the inventory is measurable or not.

pm-harshad-mane replied:
bidder's onBidViewable handler will trigger only when this module is present. 
We can add a boolean flag in PBJS name-space to let bidders know if the module is added or not.

pm-harshad-mane replied:
Hello @mike-chowla 

I have made changes in module config, now we need to set enabled flag to true to enable the module.
Bidders can check this flag to know whether the module is present or not.
```
pbjs.setConfig({
        bidViewability: {
            enabled: true
        }
    });
```

mike-chowla reviewed (approved):
LGTM

pm-harshad-mane commented:
Thank You @mike-chowla for reviewing

pm-harshad-mane commented:
@bretg, the code changes are reviewed by Jason and Mike, can we merge the changes to the main branch?

josepowera commented:
We just tryed to DOWNLOAD on https://docs.prebid.org/download.html

with modules selected:
```
modules[]: adoceanBidAdapter
modules[]: appnexusBidAdapter
modules[]: pubmaticBidAdapter
modules[]: smartadserverBidAdapter
modules[]: currency
modules[]: priceFloors
modules[]: currency
modules[]: sharedIdSystem
version: 4.27.0


```
Download is OK

-----------------
However, when we add bidViewable

```
modules[]: adoceanBidAdapter
modules[]: appnexusBidAdapter
modules[]: pubmaticBidAdapter
modules[]: smartadserverBidAdapter
modules[]: bidViewable
modules[]: currency
modules[]: priceFloors
modules[]: currency
modules[]: sharedIdSystem
version: 4.27.0
```


Download fails ({"error":"Prebid file not built properly","requestId":"fe72807a-4c3e-425b-bed7-a9a59aba7648"})

jsnellbaker commented:
Hi @josepowera 

Thanks for reporting this issue with the builds from the Download page of the prebid.org site.  I took a look and put together a fix (see link above).  When it's merged, it should address the issue.  

I will follow-up so you know when you can try again.

jsnellbaker commented:
Hi @josepowera 

The fix was merged and it seems to be working now when this module is selected.

Please let us know if you continue to see any issues.  Thanks.

bretg commented:
Original docs PR https://github.com/prebid/prebid.github.io/pull/2659 didn't add the module to the index. Fixed with https://github.com/prebid/prebid.github.io/pull/2902.
